### ISSUE TITLE
textSize() causes a memory leak in Firefox and Safari

### ISSUE THREAD

**Issue by kevinb7** [processing-js/processing-js#196]
The following code causes Firefox to use loads of memory (20GB before I killed the process):
```
var draw = function() {
    textSize(random(10, 30));
};
```
Each call to `textSize` causes a new `PFont` instance which computes font metrics in the constructor.  Computing font metrics involves creating a new canvas, rendering some text, and then inspecting all of the pixels.  This is computational expensive and uses a lot of memory.  Would it be possible to have metrics computed lazily or perhaps use something like opentype.js to compute metrics once and then scale them for different sizes?

To see the memory use skyrocket, open this like in Firefox: https://www.khanacademy.org/computer-programming/spin-off-of-contest-100-organ-ic-code/6277801210675200.

The same thing happens in Safari but it's a little harder to see in Activity Monitor because it shows up with a white lego block icon instead of the blue compass.

**Comment by kevinb7** [comment]
I retested Chrome and although Activity Monitor doesn't show memory spiking in association with any of the Chrome processes, the overall "memory used" for the system did shoot up from 3GB to 15GB in a matter of minutes while running this program in Chrome.  I'm not sure what voodoo Google is pull to be able to hide memory use from the Activity Monitor like that.

**Comment by Pomax** [comment]
that's pretty bad, if the family and size remain the same, the PFont should be fetched from cache... marking as bug,

**Comment by kevinb7** [comment]
The family is the same, but the size is random so it's not being fetched from the cache.

**Comment by Pomax** [comment]
yeah, updated my response. `random()` for fontsizes may look convenient, but it's also pretty weird - putting a `nf` in there to make sure you're not setting sizes to 1 millionth point size difference would be my first recommendation (even as plain Processing code)

**Comment by kevinb7** [comment]
I'll try updating `textSize` to round to the nearest tenth.  I guess keeping it in the cache is why memory never goes down.  It might be good to remove older entries in the cache after it reaches a certain size.

**Comment by Pomax** [comment]
yeah, although cache invalidation is basically the hardest problem in computing: no matter which cache limit I pick, someone is going to going to argue it's *just* the wrong value, so it seems more worth leaving it as is, and having this pop up every now and again. Simply using `round()` might also make sense, since you're randomizing between 10 and 30, which is a pretty huge range (given normal sketch sizes, at least. All the times I used randomized font sizes, the fractions didn't end up actually mattering)

**Comment by kevinb7** [comment]
> someone is going to going to argue it's just the wrong value

So true.

I should've mentioned that I didn't actual write the code.  I'm trying to protect Khan Academy users from themselves.  Many of our users are novice developers and so they don't necessarily know why constructing a program in a particular way makes the browser crash.  We have our own fork of processing.js which I can update so that `textSize` does the round.  If the rounding behaviour is something you think your users would appreciate I can open pull request.  If not, feel free to close this issue.

**Comment by Pomax** [comment]
good point. textSize can be indeed be made to do the rounding automatically; I'm currently at CascadiaJS but I might be able to get to it this weekend (or if you want to beat me to it, PRs are of course welcome, although in no way necessary =)

**Comment by kevinb7** [comment]
We got some more reports of our users running into this issue, not with random sizes, but ever increasing sizes.  One of our challenges has the font size ever increasing as part of an animation.  Using the rounding trick won't work in this case.  I was thinking of solving this using an LRU cache.  Thoughts?

**Comment by Pomax** [comment]
hm. we could add a @Pjs predirective that turns off font caching I suppose.

**Comment by kevinb7** [comment]
Are there some existing predirectives that you could point me at?

**Comment by Pomax** [comment]
there's four at the moment, listed fist on http://processingjs.org/reference, there's one to preload a font resource but not for controlling font caching policy.
